torch/csrc/jit/pybind_utils.h

```diff
@@ -88,82 +97,146 @@ inline c10::optional<TypePtr> unifyOrInitializeType(
   return unifyTypes(accum, unify);
 }
 
-inline TypedIValue toTypedIValue(py::handle input) {
+MatchTypeReturn tryToInferContainerType(py::handle input);
+
+// Try to infer the type of a Python object
+// The type cannot be inferred if:
+//   input is a None
+//   input is an empty container (list, dict, tuple)
+//   input is an empty container
+//   input is an dict with more than one type of key or value
+//   input is an list with more than one type of element
+inline MatchTypeReturn tryToInferType(py::handle input) {
+  // Try tensor types
   if (THPVariable_Check(input.ptr())) {
-    auto ten = py::cast<at::Tensor>(input);
-    if (ten.is_sparse()) {
-      AT_ERROR("sparse tensors not supported");
+    auto tensor = py::cast<at::Tensor>(input);
+    if (tensor.is_sparse()) {
+      return MatchTypeReturn("Sparse tensors not supported");
+    }
+    if (tensor.is_mkldnn()) {
+      // mkldnn tensor as opaque tensor doesn't have strides, so we can
+      // not create a CompleteTensorType
+      return MatchTypeReturn(DimensionedTensorType::create(tensor));
+    }
+    return MatchTypeReturn(CompleteTensorType::create(tensor));
```

Comment:
Can you add a test case where the element  / list element is a tensor type, and/or make sure we are not overspecializing here? When we try to use specialized tensor types it messes up things in a lot of places.
e.g. 
```
self.list = [Float(*, *)]
dim_tensor = self.list[0]
dim_tensor = torch.tensor()
```
Here, torch.tensor is not a subtype of Float(*, *) so it would throw

Reply:
We should do that in a follow up, this is just a refactor and behavior shouldn't change here